crates/sui-core/src/validator_client_monitor/monitor.rs

```diff
@@ -0,0 +1,300 @@
+        let authority_agg = self.authority_aggregator.load();
+        let committee = &authority_agg.committee;
+
+        // Calculate scores for all validators
+        let score_map = self
+            .client_stats
+            .read()
+            .get_all_validator_stats(committee, &authority_agg.validator_display_names);
+
+        for (validator, score) in score_map.iter() {
+            debug!("Validator {}: score {}", validator, score);
+            self.metrics
+                .performance_score
+                .with_label_values(&[&validator.concise().to_string()])
+                .set(*score);
+        }
+
+        // Cache the scores
+        *self.cached_scores.write() = Some(score_map);
+    }
+
+    /// Record client-observed interaction result with a validator.
+    ///
+    /// Records operation results including success/failure status and latency
+    /// from the client's perspective. Updates both Prometheus metrics and
+    /// internal client statistics. This is the primary interface for the
+    /// TransactionDriver to report client-observed validator interactions.
+    /// TODO: Consider adding a byzantine flag to the feedback.
+    pub fn record_interaction_result(&self, feedback: OperationFeedback) {
+        let validator_str = feedback.validator.clone();
+        let operation_str = match feedback.operation {
+            OperationType::Submit => "submit",
+            OperationType::Effects => "effects",
+            OperationType::HealthCheck => "health_check",
+        };
+
+        // Update metrics
+        if let Some(actual_latency) = feedback.latency {
+            self.metrics
+                .observed_latency
+                .with_label_values(&[&validator_str, operation_str])
+                .observe(actual_latency.as_secs_f64());
+        }
+
+        if feedback.success {
+            self.metrics
+                .operation_success
+                .with_label_values(&[&validator_str, operation_str])
+                .inc();
+        } else {
+            self.metrics
+                .operation_failure
+                .with_label_values(&[&validator_str, operation_str])
+                .inc();
+        }
+
+        // Update client stats
+        let mut client_stats = self.client_stats.write();
+        client_stats.record_interaction_result(feedback, &self.metrics);
+    }
+
+    /// Handle epoch changes by cleaning up stale validator data.
+    ///
+    /// Called when the network transitions to a new epoch. Removes client-observed
+    /// data for validators that are no longer in the active committee to prevent
+    /// memory leaks and ensure scores are only calculated for current validators.
+    /// Also clears cached scores to ensure fresh calculations for the new epoch.
+    pub fn on_epoch_change(&self, epoch: EpochId) {
+        let authority_agg = self.authority_aggregator.load();
+        if authority_agg.committee.epoch == epoch {
+            return;
+        }
```

Comment:
My understanding is that the `authority_aggregator` as it's behind an ArcSwap is getting updated async compared to when the `on_epoch_change` is called. Isn't possible that the `authority_aggregator` gets updated before we manage to call the `on_epoch_change` , so we never end up cleaning the `refresh_validator_set` due to an early return? Am I missing something?

Reply:
Good catch! Will simplify this.